File: Angular/src/app/core/services/reports.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiClientService } from './api-client.service';
import type {
  CostKpisDto,
  CostPerformancePointDto,
  CostVarianceByCostCodeDto,
  EarnedValuePointDto,
  PortfolioKpisDto,
  ProjectHealthDistributionDto,
  UpcomingMilestoneDto,
} from '../models/api.models';

@Injectable({ providedIn: 'root' })
export class ReportsService {
  private api = inject(ApiClientService);

  getPortfolioKpis(): Observable<PortfolioKpisDto> {
    return this.api.getJson<PortfolioKpisDto>('reports/portfolio-kpis');
  }

  getCostPerformanceTrend(months = 12): Observable<CostPerformancePointDto[]> {
    return this.api.getJson<CostPerformancePointDto[]>('reports/cost-performance-trend', { months });
  }

  getProjectHealthDistribution(): Observable<ProjectHealthDistributionDto> {
    return this.api.getJson<ProjectHealthDistributionDto>('reports/project-health-distribution');
  }

  getUpcomingMilestones(days = 30, limit = 20): Observable<UpcomingMilestoneDto[]> {
    return this.api.getJson<UpcomingMilestoneDto[]>('reports/upcoming-milestones', { days, limit });
  }

  getCostKpis(): Observable<CostKpisDto> {
    return this.api.getJson<CostKpisDto>('reports/cost-kpis');
  }

  getEarnedValueTrend(months = 12): Observable<EarnedValuePointDto[]> {
    return this.api.getJson<EarnedValuePointDto[]>('reports/earned-value-trend', { months });
  }

  getCostVarianceByCostCode(): Observable<CostVarianceByCostCodeDto[]> {
    return this.api.getJson<CostVarianceByCostCodeDto[]>('reports/cost-variance-by-cost-code');
  }
}

File: Blazor/Core/Services/ReportsService.cs
using Construction.Core.DTOs;

namespace Construction.Blazor.Core.Services;

public class ReportsService(ApiClient api)
{
    public Task<PortfolioKpisDto> GetPortfolioKpisAsync() =>
        api.GetJsonAsync<PortfolioKpisDto>("reports/portfolio-kpis");

    public Task<List<CostPerformancePointDto>> GetCostPerformanceTrendAsync(int months = 12) =>
        api.GetJsonAsync<List<CostPerformancePointDto>>("reports/cost-performance-trend", new Dictionary<string, object?> { ["months"] = months });

    public Task<ProjectHealthDistributionDto> GetProjectHealthDistributionAsync() =>
        api.GetJsonAsync<ProjectHealthDistributionDto>("reports/project-health-distribution");

    public Task<List<UpcomingMilestoneDto>> GetUpcomingMilestonesAsync(int days = 30, int limit = 20) =>
        api.GetJsonAsync<List<UpcomingMilestoneDto>>("reports/upcoming-milestones", new Dictionary<string, object?> { ["days"] = days, ["limit"] = limit });

    public Task<CostKpisDto> GetCostKpisAsync() =>
        api.GetJsonAsync<CostKpisDto>("reports/cost-kpis");

    public Task<List<EarnedValuePointDto>> GetEarnedValueTrendAsync(int months = 12) =>
        api.GetJsonAsync<List<EarnedValuePointDto>>("reports/earned-value-trend", new Dictionary<string, object?> { ["months"] = months });

    public Task<List<CostVarianceByCostCodeDto>> GetCostVarianceByCostCodeAsync() =>
        api.GetJsonAsync<List<CostVarianceByCostCodeDto>>("reports/cost-variance-by-cost-code");
}
